### PR TITLE
Add Hspec Tests

### PR DESCRIPTION
**Pull request by amitaibu**
As part of https://github.com/digitallyinduced/ihp/pull/1126 I'm trying to create tests that are more meaningful, and easier to copy/ paste or see in action.

Still a few issues going on, even after I've set the rev to be https://github.com/digitallyinduced/ihp/commit/f12733f8b0f6e9725d93b67051f36c896bc55edf

image

### CODE REVIEW

File: Test/Controller/PostsSpec.hs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+tests :: Spec
+tests = beforeAll (mockContext WebApplication config) do
+    describe "User controller" $ do
+        it "has no existing posts" $ withContext do
+            users <- query @Post |> fetch
+            users `shouldBe` []
+
+        it "responds with some content" $ withContext do
+            content <- mockActionResponse NewPostAction
+            isInfixOf "You can use Markdown here" (cs content) `shouldBe` True
```
Comment (amitaibu) [inline Test/Controller/PostsSpec.hs]: Two things I'd be happy to know, and to add here:

1. How to find some text/elements in HTML . For example in PHP I can do `$this->assertSession()->elementExists('css', '.some-class a');` or `$session->elementTextContains('css', '.my-selector', 'Some text');`

2. How could I debug this. This is, is there a way of printing `content` to the console, so I could see what the output is?

Comment (mpscholten) [reply Test/Controller/PostsSpec.hs]: [Just pushed some improvements to the testing infrastructure.](https://github.com/digitallyinduced/ihp/commit/9c53275aeb1581718aba76e72e59cf16ed515693) It's on the IHP repo inside the `testing-improvements` branch https://github.com/digitallyinduced/ihp/tree/testing-improvements

You can test it in this repo using this:
```bash
cd ihp-blog-example-app
git clone git@github.com:digitallyinduced/ihp.git IHP
cd IHP
git checkout testing-improvements
```

(I needed to add `async-pool` to the projects default.nix temporarly to get it working).

Mainly it now gives every test it's own dedicated empty database (previously it was using the local dev database).

Here's my commit (I couldn't push it onto the PR branch) https://github.com/digitallyinduced/ihp-blog-example-app/commit/68e606e6679f76eec4fecf7d15379b7a96ce201e Maybe you can cherry-pick that into this PR?

> How to find some text/elements in HTML . For example in PHP I can do $this->assertSession()->elementExists('css', '.some-class a'); or $session->elementTextContains('css', '.my-selector', 'Some text');

Still to be implemented.


> How could I debug this. This is, is there a way of printing content to the console, so I could see what the output is?

With the new version you could do something like this:

```haskell
response <- callAction SomeAction
body <- responseBody response
putStrLn (cs body)
```

**Comment by amitaibu** [comment]
Thanks, it's much nicer with this new API. Couple of notes:

1. Like you I had to add `async-pool` manually, I guess that would not be needed once merged.
2. There's some excessive info, and a few unclear `Nothing`s that are being printed - that I think could be removed.

Selection_999(342)

**Comment by mpscholten** [comment]
Just fixed the output 👍 

image

**Comment by amitaibu** [comment]
@mpscholten ready for review

**Review by mpscholten — APPROVED** [review]
Looks good, thanks! :)

**Pull request merged**